web/tests/stream_data.test.js

```diff
@@ -168,19 +165,37 @@ test("basics", () => {
 
     // "new" correct url formats
     assert.equal(stream_data.slug_to_stream_id("2-social"), 2);
+    assert.equal(hash_util.decode_operand("channel", "2-social"), "2");
+
     assert.equal(stream_data.slug_to_stream_id("2"), 2);
+    assert.equal(hash_util.decode_operand("channel", "2"), "2");
+
     // we still get 2 because it's a valid stream id
     assert.equal(stream_data.slug_to_stream_id("2-whatever"), 2);
-    // invalid stream id
-    assert.equal(stream_data.slug_to_stream_id("999-social"), undefined);
-    // legacy
+    assert.equal(stream_data.slug_to_stream_id("2-"), 2);
+
+    // legacy, we recognize "social" as a valid channel name
     assert.equal(stream_data.slug_to_stream_id("social"), 2);
+    assert.equal(hash_util.decode_operand("channel", "social"), "2");
+
+    // These aren't prepended with valid ids nor valid channel names. We
+    // don't get any stream id from the slug, and the decoded operand (the
+    // only caller of `slug_to_stream_id`) returns an empty string (which we
+    // don't display anywhere, since the channel is invalid).
+    assert.equal(stream_data.slug_to_stream_id("999-social"), undefined);
+    assert.equal(hash_util.decode_operand("channel", "999-social"), "");
+
+    assert.equal(stream_data.slug_to_stream_id("99-whatever"), undefined);
+    assert.equal(hash_util.decode_operand("channel", "99-whatever"), "");
 
-    // invalid formats
     assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);
```

Comment:
I'm not seeing a test that reflects the outcome of this thread above:
https://github.com/zulip/zulip/pull/31299#discussion_r1722289142

This line where the answer is "undefined" comes from the recent 05e58de3f5f507b23d287e678e96879937ad29b7 (#31575), which was split out from this PR. This format is still valid and should still work, when a stream named "25-or-6-to-4" exists (well, is known) and one with ID 25 doesn't.

Reply:
(I think the new implementation does handle that case correctly — there should just be a test showing that.)

Reply:
There are these two tests

```js
    assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);
    assert.equal(hash_util.decode_operand("channel", "25-or-6-to-4"), "");
```

The part where we `return operand;` because there's no stream id is a function of `hash_util.decode_operand`, not anything in `stream_data`. Did you want a test added for `hash_util` for this? Or something else?

Reply:
Right, those are the test lines I was commenting on :slightly_smiling_face: (or one of them is).

This format is still valid. That means that `slug_to_stream_id` should be able to return an ID for it.

Currently I think we don't have any tests that would fail if this bit of the implementation:
```js
    const newFormatStreamId = match ? Number.parseInt(match[1]!, 10) : null;
    if (newFormatStreamId !== null && stream_info.get(newFormatStreamId)) {
        return newFormatStreamId;
    }
```

were edited to instead look like:
```js
    if (match) {
        // The slug is in the new format.
        const streamId = Number.parseInt(match[1]!, 10);
        if (!stream_info.get(streamId)) {
            // The stream ID isn't for a known stream.
            return null;
        }
        return streamId;
    }
```

Logic like that would look pretty reasonable on the surface, but would break old links where the stream name happens to have the same shape as a new-style slug. So we should have tests that would detect that breakage.

As I said in the linked thread, that's what the "311-help", "311-", and "311" cases in these mobile tests are testing:
https://github.com/zulip/zulip-mobile/blob/2217c858e207f9f092651dd853051843c3f04422/src/utils/__tests__/internalLinks-test.js#L333-L354

Reply:
I believe that's what these tests are testing (earlier in the file)? Sorry if I'm missing something still.

```js
    assert.equal(stream_data.slug_to_stream_id("2-social"), 2);
    assert.equal(stream_data.slug_to_stream_id("2"), 2);
    assert.equal(stream_data.slug_to_stream_id("2-whatever"), 2);
    assert.equal(stream_data.slug_to_stream_id("2-"), 2);
```

Reply:
So you're looking for a test that has a stream with a name that starts with a number?

Like maybe after `assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);` adding a stream with the name `"25-or-6-to-4"` and confirming that it identifies that stream. Am I understanding now?

Reply:
Right, exactly.

Then as a supplement to that, it would also be good to have a stream with a numeric name like "25" (or some other number), and check that `slug_to_stream_id` returns the ID of that stream (not undefined, and not the number 25).

And as a further case (or rather pair of cases), check that if the name is like either "25" or "25-or-6-to-4" and there's *both* a stream with ID 25 and a different stream that has the whole slug as its name, the one with the ID wins.

Reply:
Ok great, that makes sense. Here's a PR with that: #31677